### PR TITLE
Mime type components schema

### PR DESCRIPTION
**Pull request by MyDigitalLife**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #2415 #2502
| License       | MIT
| Doc PR        | 

This is a PR for being able to have different body for mime types.

After these tests are fixed there should be a different body schema created for every mime type. The basics of this can be laid down in this pull request and adding the correct bodies can be done later as long as there is a default body that falls back to how things work now.

Any input in how this can be improved are welcome. 

- [x] Extract  JSON Schema formatting to its own classes
- [x] Load schema formatters using tags: https://symfony.com/doc/current/service_container/tags.html
- [x] Improve formatters to comply with at least 2 different standards (other can be added later on or in a other pull requests)
- [x] Fix tests

### CODE REVIEW

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 1]
```diff
@@ -604,14 +707,37 @@ private function getDefinitionSchema(bool $v3, string $resourceClass, ResourceMe
         }
 
         $options = isset($serializerContext[AbstractNormalizer::GROUPS]) ? ['serializer_groups' => $serializerContext[AbstractNormalizer::GROUPS]] : [];
+
+        if ($v3) {
+            if ('application/vnd.api+json' === $mimeType) {
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Cab we extract the code generating the JSON Schema corresponding to every supported formats in specific classes. It will make this class less complex (it is already to big and complex), and will allow users to register their on JSON Schema generators for their custom formats.

Comment (MyDigitalLife) [reply src/Swagger/Serializer/DocumentationNormalizer.php]: Yeah, that's on my todo list. I'll add a todo list to the so we know what needs to happen.

Comment (MyDigitalLife) [reply src/Swagger/Serializer/DocumentationNormalizer.php]: I added some functionality to do this, its not finished still needs to get the formatters from the config and create more functions for more complex setups. Its a start, just let me know if its going in the correct direction.

**Comment by MyDigitalLife** [comment]
Alright, looks like this is mostly done. I would love some feedback. There are a few loose ends to tie up (php-cs-fixer and a few few test that still fail), but I would love some feedback. Mostly about the following points:

- Fix a one to many relationship in the JsonApiFormatter, there aren't any current tests that have this relationship in for the docs. I would love some help writing this one so I can finish the check when a relationship is a one to many or a many to one. ATM I don't have the knowledge to mock the resources I need for this. 
- Fix the broken tests. There are 3 tests that re broken and I have no idea why, you can find those [here](https://circleci.com/gh/api-platform/core/11412?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). I would love some help with fixing these issues. 
- I think I'm using some design pattern name wrong, I'm calling the class i use to get the correct formatter a Factory but I'm not use that correct. 
- General improvements I can make to the code or maybe some extra tests i can write. 

cc @dunglas @teohhanhui 

Looking forward to your feedback and to get the code to the point it can be merged and start writing more formatters.

**Comment by antograssiot** [comment]
The failing tests are because you change the path of references in component's schemas.

In [features/openapi/docs.feature](https://github.com/api-platform/core/blob/bc0806be383a6fc6820a1230bbddd3ea29542ac7/features/openapi/docs.feature#L19) we are checking in the generated OpenAPI for an existing class being defined using [SwaggerContext ::getClassInfo()](https://github.com/api-platform/core/blob/bc0806be383a6fc6820a1230bbddd3ea29542ac7/features/bootstrap/SwaggerContext.php#L178)

For OpenAPI v3, classes were expected in `#/components/schemas/classname` but now there are stored in `#/components/schemas/mimetype/classname` so you need to update this function

**Comment by MyDigitalLife** [comment]
Thanks for the feedback.

I still need to write a test and implement a one to many relationship. I would love some help with this as I haven't figured out how to mock that in the correct tests.

File: src/Swagger/SchemaFormatter/JsonApiSchemaFormatter.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            ],
+        ];
+    }
+
+    public function setProperty(
+        \ArrayObject $definitionSchema,
+        $normalizedPropertyName,
+        \ArrayObject $property,
+        PropertyMetadata $propertyMetadata
+    ): void {
```
Comment (soyuka) [inline src/Swagger/SchemaFormatter/JsonApiSchemaFormatter.php]: We usually inline these.

File: src/Swagger/SchemaFormatter/JsonApiSchemaFormatter.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            ];
+            // @todo: Fix one to many statement.
+//            if (false) {
+//                $data = [
+//                    'type' => 'object',
+//                    'properties' => [
+//                        $data,
+//                    ],
+//                ];
+//            }
```
Comment (soyuka) [inline src/Swagger/SchemaFormatter/JsonApiSchemaFormatter.php]: We may leave the todo comment but not the commented code

File: src/Swagger/SchemaFormatter/JsonApiSchemaFormatter.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                                        'type' => 'string',
+                                    ],
+                                ],
+                            ],
+                            'data' => $data,
+                        ],
+                    ],
+                ],
+            ];
+        } else {
```
Comment (soyuka) [inline src/Swagger/SchemaFormatter/JsonApiSchemaFormatter.php]: return early to avoid the else?

File: src/Swagger/SchemaFormatter/JsonSchemaFormatter.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        return [];
+    }
+
+    public function setProperty(
+        \ArrayObject $definitionSchema,
+        $normalizedPropertyName,
+        \ArrayObject $property,
+        PropertyMetadata $propertyMetadata
+    ): void {
```
Comment (soyuka) [inline src/Swagger/SchemaFormatter/JsonSchemaFormatter.php]: inline this?

File: src/Swagger/SchemaFormatter/SchemaFormatterProvider.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Swagger\SchemaFormatter;
+
+use ApiPlatform\Core\Exception\FormatterNotFoundException;
+
+class SchemaFormatterProvider
+{
+    /** @var SchemaFormatterInterface[] */
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/SchemaFormatterProvider.php]: Move it to the contructors's doc block

File: src/Swagger/SchemaFormatter/SchemaFormatterProvider.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        $this->formatters = $formatters;
+    }
+
+    /**
+     * @param string $mimeType
+     *
+     * @return SchemaFormatterInterface
+     */
+    public function getFormatter($mimeType)
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/SchemaFormatterProvider.php]:
```suggestion
    public function getFormatter(string $mimeType): SchemaFormatterInterface
```

File: src/Swagger/SchemaFormatter/SchemaFormatterProvider.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+{
+    /** @var SchemaFormatterInterface[] */
+    private $formatters;
+
+    public function __construct(/* iterable */ $formatters)
+    {
+        $this->formatters = $formatters;
+    }
+
+    /**
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/SchemaFormatterProvider.php]: DocBlock to be removed, use typehints instead

File: src/Swagger/SchemaFormatter/SchemaFormatterProvider.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Swagger\SchemaFormatter;
+
+use ApiPlatform\Core\Exception\FormatterNotFoundException;
+
+class SchemaFormatterProvider
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/SchemaFormatterProvider.php]: For consistency with the rest of the code base (see DataProvider / DataPersister for instance), you should rename this class `ChainSchemaFormatter` and make it an implementation of `SchemaFormatterInterface`. Then only inject `SchemaFormatterInterface` in the normalizer.

Also:
```suggestion
final class SchemaFormatterProvider
```

File: src/Swagger/SchemaFormatter/SchemaFormatterProvider.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public function getFormatter($mimeType)
+    {
+        foreach ($this->formatters as $formatter) {
+            if ($formatter->supports($mimeType)) {
+                return $formatter;
+            }
+        }
+
+        throw new FormatterNotFoundException(sprintf('Formatter for mimetype "%s" not supported', $mimeType));
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/SchemaFormatterProvider.php]:
```suggestion
        throw new FormatterNotFoundException(sprintf('No formatter supporting the "%s" MIME type is available.', $mimeType));
```

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 11]
```diff
@@ -87,11 +88,12 @@ final class DocumentationNormalizer implements NormalizerInterface, CacheableSup
...
+    public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFactory, PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, ResourceClassResolverInterface $resourceClassResolver, OperationMethodResolverInterface $operationMethodResolver, OperationPathResolverInterface $operationPathResolver, UrlGeneratorInterface $urlGenerator = null, $filterLocator = null, NameConverterInterface $nameConverter = null, bool $oauthEnabled = false, string $oauthType = '', string $oauthFlow = '', string $oauthTokenUrl = '', string $oauthAuthorizationUrl = '', array $oauthScopes = [], array $apiKeys = [], SubresourceOperationFactoryInterface $subresourceOperationFactory = null, bool $paginationEnabled = true, string $paginationPageParameterName = 'page', bool $clientItemsPerPage = false, string $itemsPerPageParameterName = 'itemsPerPage', OperationAwareFormatsProviderInterface $formatsProvider = null, bool $paginationClientEnabled = false, string $paginationClientEnabledParameterName = 'pagination', array $defaultContext = [], SchemaFormatterProvider $schemaFormatterFactory = null)
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: `SchemaFormatterInterface $schemaFormatter` instead of `SchemaFormatterProvider` (see my previous comment)

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 12]
```diff
@@ -116,14 +118,14 @@ public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFa
         $this->paginationEnabled = $paginationEnabled;
         $this->paginationPageParameterName = $paginationPageParameterName;
         $this->apiKeys = $apiKeys;
-        $this->subresourceOperationFactory = $subresourceOperationFactory;
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Should be reverted I guess

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 13]
```diff
@@ -241,9 +243,17 @@ private function updateGetOperation(bool $v3, \ArrayObject $pathOperation, array
         $serializerContext = $this->getSerializerContext($operationType, false, $resourceMetadata, $operationName);
 
         $responseDefinitionKey = false;
+        $responseDefinitionKeys = [];
         $outputMetadata = $resourceMetadata->getTypedOperationAttribute($operationType, $operationName, 'output', ['class' => $resourceClass], true);
         if (null !== $outputClass = $outputMetadata['class'] ?? null) {
-            $responseDefinitionKey = $this->getDefinition($v3, $definitions, $resourceMetadata, $resourceClass, $outputClass, $serializerContext);
+            if ($v3) {
+                foreach ($mimeTypes as $mimeType) {
+                    $responseDefinitionKeys[$mimeType] = $this->getDefinition($v3, $definitions, $resourceMetadata, $resourceClass,
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: What about:

* Renaming `SchemaFormatterInterface` in `DefinititionNormalizerInterface`
* Adding only a simple single method to this interface such as `SchemaFormatterInterface::getDefinition(string $resourceClass, string $operationType, string $operationName = null, string $inputOrOutput, array $serializerContext = [], string $mimeType): array` (other arguments such as outputClass and  can be generated using the factory services)

It would simplify the code a lot

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 14]
```diff
@@ -256,16 +266,21 @@ private function updateGetOperation(bool $v3, \ArrayObject $pathOperation, array
...
-            if ($responseDefinitionKey) {
-                if ($v3) {
-                    $successResponse['content'] = array_fill_keys($mimeTypes, [
-                        'schema' => [
-                            'type' => 'array',
-                            'items' => ['$ref' => sprintf('#/components/schemas/%s', $responseDefinitionKey)],
-                        ],
-                    ]);
-                } else {
+            if ($v3) {
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Is this change necessary? It generates the same output AFAIU

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 15]
```diff
@@ -286,10 +301,19 @@ private function updateGetOperation(bool $v3, \ArrayObject $pathOperation, array
         $pathOperation = $this->addItemOperationParameters($v3, $pathOperation);
 
         $successResponse = ['description' => sprintf('%s resource response', $resourceShortName)];
-        if ($responseDefinitionKey) {
-            if ($v3) {
-                $successResponse['content'] = array_fill_keys($mimeTypes, ['schema' => ['$ref' => sprintf('#/components/schemas/%s', $responseDefinitionKey)]]);
-            } else {
+
+        if ($v3) {
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Same here

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 16]
```diff
@@ -408,7 +446,16 @@ private function getSubresourceResponse(bool $v3, $mimeTypes, bool $collection,
             ];
 
             if ($v3) {
-                $okResponse['content'] = array_fill_keys($mimeTypes, ['schema' => ['$ref' => sprintf('#/components/schemas/%s', $definitionKey)]]);
+                foreach ($mimeTypes as $mimeType) {
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Same here. Try to reduce the diff size as much as possible to ease the review process

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 17]
```diff
@@ -601,14 +714,30 @@ private function getDefinitionSchema(bool $v3, string $resourceClass, ResourceMe
         }
 
         $options = isset($serializerContext[AbstractNormalizer::GROUPS]) ? ['serializer_groups' => $serializerContext[AbstractNormalizer::GROUPS]] : [];
+
+        $schemaFormatter = null;
+        if ($v3) {
+            try {
+                $schemaFormatter = $this->schemaFormatterFactory->getFormatter($mimeType);
+            } catch (\Exception $e) {
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: The default formatter's `supports` method should just always return `true` instead of thowing. It will allow to remove this hack.

File: src/Swagger/SchemaFormatter/JsonSchemaFormatter.php [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Swagger\SchemaFormatter;
+
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+
+class JsonSchemaFormatter implements SchemaFormatterInterface
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/JsonSchemaFormatter.php]: Should be renamed `DefaultSchemaFormatter`, cause it can it's the default one, also used for XML and so on.

File: src/Swagger/SchemaFormatter/JsonSchemaFormatter.php [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public function supports(string $mimeType): bool
+    {
+        return 'application/json' === $mimeType;
```
Comment (dunglas) [inline src/Swagger/SchemaFormatter/JsonSchemaFormatter.php]: Should always return `true`, and should be registered with a low priority.

File: src/Swagger/Serializer/DocumentationNormalizer.php (hunk 12 above)
Comment (MyDigitalLife) [reply src/Swagger/Serializer/DocumentationNormalizer.php]: Actually this line is duplicate, check a few lines up. I just removed the duplicate line.

File: src/Swagger/Serializer/DocumentationNormalizer.php (hunk 13 above)
Comment (Simperfit) [reply src/Swagger/Serializer/DocumentationNormalizer.php]: this has been done
